### PR TITLE
Event rule trigger For BlockAppended and TxConfirmed

### PR DESCRIPTION
**Pull request by hvoid-build-block**
# PR Details

## Description
Add event trigger in blockchainUpdaterImpl to trigger BlockAppended event and TxConfirmed event.

## Related Issue
add new feature

## How Has This Been Tested
unit tests passed

## Types of changes

- [x]   New feature (non-breaking change which adds functionality)

## Checklist

- [x]   My code follows the code style of this project.
- [x]   I have added tests to cover my changes.
- [x]   All new and existing tests passed.

### CODE REVIEW

File: src/test/scala/vsys/settings/WebhookEventRulesSpec.scala [hunk 1]
```diff
@@ -51,6 +56,72 @@ class WebhookEventRulesSpec extends FlatSpec with Matchers {
...
+    ExcludeTypes(Seq(1, 2)).applyRule(0, mockTx_2, Set.empty) shouldBe(false)
+    ExcludeTypes(Seq.empty).applyRule(0, mockTx_1, Set.empty) shouldBe(true)
+
+    Amount(Seq(AmtWithFee(true))).applyRule(0, mockTx_1, Set.empty) shouldBe(true)
+    Amount(Seq(AmtWithFee(false))).applyRule(0, mockTx_1, Set.empty) shouldBe(true)
+    Amount(Seq.empty).applyRule(0, mockTx_1, Set.empty) shouldBe(true)
+
+    Amount(Seq(AmtWithFee(false), AmtGT(200))).applyRule(0, mockTx_2, Set.empty) shouldBe(false)
+    Amount(Seq(AmtWithFee(true), AmtGT(200))).applyRule(0, mockTx_2, Set.empty) shouldBe(true)
+    Amount(Seq(AmtGT(250))).applyRule(0, mockTx_2, Set.empty) shouldBe(false)
```
Comment (ncying) [inline src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: so boundary case should be checked

File: src/main/scala/vsys/events/EventWriterActor.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+class EventWriterActor() extends Actor with ScorexLogging {
+
+  // TO DO: Should have MQ and protect memory usage in JVM
+  // It should handle more events
+  override def receive = {
+    case BlockAppendedEvent(url, secretKey, enKey, maxSize, eventData) =>
+      eventData.map {aList =>
+        aList._3.map(addr => log.info(addr.toString))
+      }
```
Comment (ncying) [inline src/main/scala/vsys/events/EventWriterActor.scala]: may use following solution to avoid the unnamed parameter style issue
```scala
      eventData.map {(_, _, addrs) =>
        addrs.map(addr => log.info(addr.toString))
      }
```

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 3]
```diff
@@ -54,7 +67,13 @@ object WithTxs extends RuleConfigReader {
   }
 }
 
-case class WithMintingTxs(value: Boolean) extends WebhookEventRules
+case class WithMintingTxs(value: Boolean) extends WebhookEventRules {
+  override def applyRule(height: Long, tx: ProcessedTransaction, accs: Set[Address]): Boolean = {
+    if (tx.transaction.transactionType == TransactionType.MintingTransaction)
+      if (value) true else false
+    else true
+  }
```
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: This rule is applied to minting tx. When a processed transaction contains minting tx, it will always return true. If it is minting tx, it will apply the rule. I have updated this code to be neat.

[the remark this answers is not in the document]
Comment (ncying) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: let me analyze the logic here
let A: `tx.transaction.transactionType == TransactionType.MintingTransaction`
B: value

|  A  | B |Return|
|---|---|---|
| true |  true |  true |
|  true |false   |false   |
|   false|  false |  true |
|false| true| ture|

The only case return false is A true B false, => !(A && !B) => !A || B, right?

[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: good catch~~

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 4]
```diff
@@ -64,7 +83,16 @@ object WithMintingTxs extends RuleConfigReader {
...
 }
 
-case class RelatedAccs(value: Seq[String]) extends WebhookEventRules
+case class RelatedAccs(value: Seq[String]) extends WebhookEventRules {
+  override def applyRule(height: Long, tx: ProcessedTransaction, accs: Set[Address]): Boolean = {
+    val addrInTx = accs.toSeq.map(_.toString)
+    if(value.nonEmpty) {
+      value.intersect(addrInTx).nonEmpty
+    } else {
+      addrInTx.isEmpty
```
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: when user sets the config to have empty relatedAccs, the system will eliminate all txs. If user does not set such rule, the system will pass all txs.

File: src/main/scala/vsys/utils/TransactionHelper.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import vsys.blockchain.transaction.lease._
+import vsys.blockchain.transaction._
+
+object TransactionHelper {
+
+  def extractAmtFee(tx: ProcessedTransaction): (Long, Long) = {
+    tx.transaction match {
+      case p: PaymentTransaction => (p.amount, p.transactionFee)
+      case dbTx: DbPutTransaction => (0, dbTx.transactionFee)
+      case lct: LeaseCancelTransaction => (0, lct.transactionFee)
```
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/utils/TransactionHelper.scala]: The compiler will pop up error if I write it this way. I may use 'NonFeeTransaction' trait and 'AmountInvolved" trait to refactor it.

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 6]
```diff
@@ -94,7 +130,28 @@ object ExcludeTypes extends RuleConfigReader {
...
+      // not withFee
+      value.filterNot(_.isInstanceOf[AmtWithFee]).foldLeft(true)((accum, rule) =>
+        rule.applyRule(amt, tx, accs) && accum)
+    }
+  }
+}
+
+case class AmtGT(value: Long) extends WebhookEventRules {
+  override def applyRule(valueIn: Long, tx: ProcessedTransaction, accs: Set[Address]): Boolean = {
+    valueIn >= value
```
Comment (ncying) [inline src/main/scala/vsys/settings/WebhookEventRules.scala]: @hvoid-build-block still miss the comment here

File: src/main/scala/vsys/Application.scala [hunk 7]
```diff
@@ -46,7 +46,7 @@ class Application(val actorSystem: ActorSystem, val settings: VsysSettings) exte
   private val db = openDB(settings.dataDirectory)
 
   private val checkpointService = new CheckpointServiceImpl(db, settings.checkpointsSettings)
-  private val (history, _, stateReader, blockchainUpdater) = StorageFactory(db, settings.blockchainSettings)
+  private val (history, _, stateReader, blockchainUpdater) = StorageFactory(db, settings.blockchainSettings, settings.eventSettings, actorSystem)
```
Comment (zl730) [inline src/main/scala/vsys/Application.scala]: May build a independent singleton out of `StrorageFactory` with  `settings.eventSettings` and `actorSystem`
this module is not a part of storage system

File: src/main/scala/vsys/events/EventTypes.scala [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package vsys.events
+
+import vsys.account.Address
+import vsys.blockchain.transaction.ProcessedTransaction
+
+trait Event
+
+case class BlockAppendedEvent(
+	url: String,
```
Comment (zl730) [inline src/main/scala/vsys/events/EventTypes.scala]: using tabs this file

File: src/main/scala/vsys/events/EventWriterActor.scala [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+class EventWriterActor() extends Actor with ScorexLogging {
+
+  // TO DO: Should have MQ and protect memory usage in JVM
+  // It should handle more events
+  override def receive = {
+    case BlockAppendedEvent(url, secretKey, enKey, maxSize, eventData) =>
+      eventData.map {aList =>
+        aList match {case (_, _, accs: Set[Address]) => accs.map(addr => log.info(addr.toString))}
+      }
```
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/events/EventWriterActor.scala]: The compiler will pop up 'missing parameter type for expanded function' when using underscore in 'log.info'.
The 'Address' type needs to have explicit 'toString' function otherwise it will have an error.

File: src/test/scala/vsys/settings/WebhookEventRulesSpec.scala [hunk 10]
```diff
@@ -51,6 +56,72 @@ class WebhookEventRulesSpec extends FlatSpec with Matchers {
...
+    ExcludeTypes(Seq(1, 2)).applyRule(0, mockTx2, Set.empty) shouldBe(false)
+    ExcludeTypes(Seq.empty).applyRule(0, mockTx1, Set.empty) shouldBe(true)
+
+    Amount(Seq(AmtWithFee(true))).applyRule(0, mockTx1, Set.empty) shouldBe(true)
+    Amount(Seq(AmtWithFee(false))).applyRule(0, mockTx1, Set.empty) shouldBe(true)
+    Amount(Seq.empty).applyRule(0, mockTx1, Set.empty) shouldBe(true)
+
+    Amount(Seq(AmtWithFee(false), AmtGT(200))).applyRule(0, mockTx2, Set.empty) shouldBe(false)
+    Amount(Seq(AmtWithFee(true), AmtGT(200))).applyRule(0, mockTx2, Set.empty) shouldBe(true)
+    Amount(Seq(AmtGT(250))).applyRule(0, mockTx2, Set.empty) shouldBe(false)
```
Comment (ncying) [inline src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: @hvoid-build-block in order to avoid the error about `AmtGT` code, please add a boundary case test here!

Comment (hvoid-build-block) [reply src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: sure~

File: src/main/scala/vsys/Application.scala (hunk 7 above)
Comment (hvoid-build-block) [reply src/main/scala/vsys/Application.scala]: we can form an `EventTrigger` with `EventWriterActor` and `EventSettings` in application.scala and pass this trigger to `StorageFactory`

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 11]
```diff
@@ -34,7 +41,11 @@ object AfterHeight extends RuleConfigReader {
   }
 }
 
-case class AfterTime(value: Long) extends WebhookEventRules
+case class AfterTime(value: Long) extends WebhookEventRules {
+  override def applyRule(height: Long, tx: ProcessedTransaction, accs: Set[Address]): Boolean = {
+    if (tx.transaction.timestamp >= value) true else false
```
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: There is another problem if we use block mint time: When system needs to handle state updated event, the trigger can only get `blockDiff` from `stateWriterImpl`, then the trigger is unable to filter txs based on block mint time.

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 12]
```diff
@@ -94,57 +128,79 @@ object ExcludeTypes extends RuleConfigReader {
...
+    } else {
+      // not withFee
+      value.filterNot(_.isInstanceOf[AmtWithFee]).foldLeft(true)((accum, rule) =>
+        rule.applyRule(amt, tx, accs) && accum)
+    }
   }
 }
 
-case class AmtLT(value: Long) extends WebhookEventRules
+object Amount extends RuleConfigReader {
```
Comment (zl730) [inline src/main/scala/vsys/settings/WebhookEventRules.scala]: Add `applyRule()` in Amount since AmtWithFee is not a rule.

Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: np

File: src/main/scala/vsys/settings/WebhookEventRules.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (hvoid-build-block) [reply src/main/scala/vsys/settings/WebhookEventRules.scala]: Yes this is dobale for state updated event. The state updated can use `persisted` variable in `BlockchainUpdaterImpl`. I will update this code.

File: src/main/scala/vsys/settings/WebhookEventRules.scala [hunk 13]
```diff
@@ -94,58 +124,47 @@ object ExcludeTypes extends RuleConfigReader {
...
+  override val field = "amount"
 
-object AmtGTE extends RuleConfigReader {
-  override val field = "amount.gte"
+  override def fromConfig(config: Config): Option[Amount] = {
+    val gtVal = getVal[Long](config, "gt").getOrElse(0L)
+    val gteVal = getVal[Long](config, "gte").getOrElse(0L)
+    val ltVal = getVal[Long](config, "lt").getOrElse(6*10^9*10^8L)
+    val lteVal = getVal[Long](config, "lte").getOrElse(6*10^9*10^8L)
```
Comment (ncying) [inline src/main/scala/vsys/settings/WebhookEventRules.scala]: Why not use `Long.MaxValue` here?

File: src/test/scala/vsys/settings/WebhookEventRulesSpec.scala [hunk 14]
```diff
@@ -41,16 +46,75 @@ class WebhookEventRulesSpec extends FlatSpec with Matchers {
...
+    Amount(0, 151L, 1000, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+    Amount(0, 250L, 1000, 1000, true).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+    Amount(0, 251L, 1000, 1000, true).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+
+    Amount(0, 0, 151, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+    Amount(0, 0, 150, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+
+    Amount(0, 0, 1000, 150, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+    Amount(0, 0, 1000, 149, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+  }
```
Comment (ncying) [inline src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: `Amount(gtVal, gteVal, ltVal, lteVal, withFeeVal)`. It is important to test all boundary cases. These will help you find the bug in the code.

For example:
if you amount is x, you should test, `Amount(x/x-1, _, _, _, _)` and `Amount(_, x/x+1, _, _, _)`

File: src/test/scala/vsys/settings/WebhookEventRulesSpec.scala [hunk 15]
```diff
@@ -41,16 +46,96 @@ class WebhookEventRulesSpec extends FlatSpec with Matchers {
...
+      Amount(0, 0, 150 + x, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+      Amount(0, 0, 1000, 150 + x, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+    }
+
+    (1 to 2).map {x =>
+      Amount(150 + x, 0, 1000, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+      Amount(0, 150 + x, 1000, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+      Amount(0, 0, 150 + x, 1000, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+      Amount(0, 0, 1000, 150 + x, false).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+    }
```
Comment (zl730) [inline src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: L101-127, need to be simplified with some code like `shoudBe(x > 0)`.

File: src/test/scala/vsys/settings/WebhookEventRulesSpec.scala [hunk 16]
```diff
@@ -41,16 +46,96 @@ class WebhookEventRulesSpec extends FlatSpec with Matchers {
...
+
+    IncludeTypes(Seq(1)).applyRule(0, blockTime1, mockTx1, Set.empty) shouldBe(false)
+    IncludeTypes(Seq(1, 2)).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(true)
+    IncludeTypes(Seq.empty).applyRule(0, blockTime1, mockTx1, Set.empty) shouldBe(false)
+
+    ExcludeTypes(Seq(1)).applyRule(0, blockTime1, mockTx1, Set.empty) shouldBe(true)
+    ExcludeTypes(Seq(1, 2)).applyRule(0, blockTime2, mockTx2, Set.empty) shouldBe(false)
+    ExcludeTypes(Seq.empty).applyRule(0, blockTime1, mockTx1, Set.empty) shouldBe(true)
+
+    (-2 to -1).map {x =>
```
Comment (zl730) [inline src/test/scala/vsys/settings/WebhookEventRulesSpec.scala]: `foreach` is better than `map` here.
> `map` is implemented by `foreach` in scala collection. So if mapping is not required, `foreach` could be more efficient.

File: src/test/scala/vsys/utils/TestTransactionHelper.scala [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    val dbTx = mock[DbPutTransaction]
+    val lct = mock[LeaseCancelTransaction]
+    val cst = mock[ContendSlotsTransaction]
+    val rst = mock[ReleaseSlotsTransaction]
+    val lt = mock[LeaseTransaction]
+    val gt = mock[GenesisTransaction]
+    val mt = mock[MintingTransaction]
+    val ecft = mock[ExecuteContractFunctionTransaction]
+    val rct = mock[RegisterContractTransaction]
+    val tx = mock[ProcessedTransaction]
```
Comment (zl730) [inline src/test/scala/vsys/utils/TestTransactionHelper.scala]: May use generators in `TransacionGen.scala` instead
